cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -812,15 +826,15 @@ class InputState {
       }
       latest_time_ = next_time;
       // If we have an active batch
-      if (++latest_ref_row_ >= (row_index_t)queue_.UnsyncFront()->num_rows()) {
+      if (++latest_ref_row_ >= (row_index_t)queue_.UnsyncFront().batch->num_rows()) {
         // hit the end of the batch, need to get the next batch if possible.
         ++batches_processed_;
         latest_ref_row_ = 0;
         have_active_batch &= !queue_.TryPop();
         if (have_active_batch) {
-          DCHECK_GT(queue_.UnsyncFront()->num_rows(), 0);  // empty batches disallowed
-          key_hasher_->Invalidate();  // batch changed - invalidate key hasher's cache
-          memo_.UpdateTime(GetTime(queue_.UnsyncFront().get(), 0));  // time changed
+          DCHECK_GT(queue_.UnsyncFront().batch->num_rows(),
+                    0);  // empty batches disallowed
+          memo_.UpdateTime(GetTime(queue_.UnsyncFront().batch.get(), 0));  // time changed
```

Comment:
Are you sure that `queue_.UnsyncFront()` here is the same as just above?

Reply:
Yes, because only the processing thread, which runs this code, is popping from the (front of the) queue.